File: frontend/src/types/interview.ts
import { PublicUser } from "./user";

export interface MockInterviewSession {
  id: string;
  user: PublicUser;
  topic: string;
  difficulty: "easy" | "medium" | "hard";
  interview_type: "technical" | "behavioral" | "system_design" | "mixed";
  status: "active" | "completed" | "abandoned";
  messages: MockInterviewMessage[];
  score: number | null;
  feedback: string | null;
  duration_seconds: number;
  created_at: string;
  updated_at: string;
}

export interface MockInterviewCreate {
  topic: string;
  difficulty: "easy" | "medium" | "hard";
  interview_type: "technical" | "behavioral" | "system_design" | "mixed";
}

export interface MockInterviewMessage {
  id: string;
  session_id: string;
  role: "interviewer" | "candidate";
  content: string;
  created_at: string;
}

export interface PeerRoom {
  id: string;
  name: string;
  host: PublicUser;
  participants: PublicUser[];
  topic: string;
  difficulty: "easy" | "medium" | "hard";
  status: "waiting" | "in_progress" | "completed";
  max_participants: number;
  problem_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface PeerRoomCreate {
  name: string;
  topic: string;
  difficulty: "easy" | "medium" | "hard";
  max_participants?: number;
  problem_id?: string;
}

File: backend/app/schemas/interview.py
import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.user import UserResponse


class MockInterviewCreate(BaseModel):
    """Request body for starting a mock interview session."""

    topic: str = Field(..., min_length=1, max_length=100)
    difficulty: str = Field(..., pattern=r"^(easy|medium|hard)$")

    model_config = ConfigDict(from_attributes=True)


class MockInterviewMessage(BaseModel):
    """A single message in a mock interview conversation."""

    role: str = Field(..., pattern=r"^(user|assistant|system)$")
    content: str

    model_config = ConfigDict(from_attributes=True)


class MockInterviewResponse(BaseModel):
    """Mock interview session representation."""

    id: uuid.UUID
    user_id: uuid.UUID
    topic: str
    difficulty: str
    status: str
    messages: list[MockInterviewMessage] = []
    feedback: Optional[str] = None
    score: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PeerRoomCreate(BaseModel):
    """Request body for creating a peer interview room."""

    title: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    topic: Optional[str] = None
    max_participants: int = Field(2, ge=2, le=10)

    model_config = ConfigDict(from_attributes=True)


class PeerRoomResponse(BaseModel):
    """Peer interview room representation."""

    id: uuid.UUID
    title: str
    description: Optional[str] = None
    topic: Optional[str] = None
    max_participants: int
    status: str
    host: UserResponse
    participants: list[UserResponse] = []
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)
